### PR TITLE
Refactor oracle migration

### PR DESCRIPTION
**Pull request by ssh24**
### Description
Refactor migration

#### Related issues
connect to https://github.com/strongloop/loopback-connector/issues/15

### CODE REVIEW

File: lib/migration.js [hunk 1]
```diff
@@ -9,35 +9,43 @@ var async = require('async');
...
    * @param {Function} cb The callback function
    */
   Oracle.prototype.getTableStatus = function(model, cb) {
-    function decoratedCallback(err, data) {
-      if (err) {
-        console.error(err);
-      }
-      if (!err) {
-        data.forEach(function(field) {
-          field.type = mapOracleDatatypes(field.type);
```
Comment (jannyHou) [inline lib/migration.js]: seems after refactor step `mapOracleDatatypes` is missing...

File: lib/migration.js [hunk 2]
```diff
@@ -9,35 +9,43 @@ var async = require('async');
...
-   * @param {String[]} [models] A model name or an array of model names. If not
-   * present, apply to all models
-   * @param {Function} [cb] The callback function
-   */
+  * Perform autoupdate for the given models
+  * @param {String[]} [models] A model name or an array of model names. If not
+  * present, apply to all models
+  * @param {Function} [cb] The callback function
+  */
   Oracle.prototype.autoupdate = function(models, cb) {
```
Comment (jannyHou) [inline lib/migration.js]: This comment might be too earlier, just a reminder, we should remove the function after refactor done :)

File: lib/migration.js [hunk 3]
```diff
@@ -437,28 +364,9 @@ function mixinMigration(Oracle) {
...
-    var columnMetadata = this.columnMetadata(model, property);
-    var colType = columnMetadata && columnMetadata.dataType;
-    if (colType) {
-      colType = colType.toUpperCase();
-    }
-    var prop = this.getModelDefinition(model).properties[property];
-    if (!prop) {
-      return null;
-    }
-    var colLength = columnMetadata && columnMetadata.dataLength || prop.length;
```
Comment (jannyHou) [inline lib/migration.js]: in base connector `colLength` can also read from `prop.limit`
```js
var colLength = columnMetadata && columnMetadata.dataLength ||
 prop.length || prop.limit;
```
IIRC this is a mysql only implementation, I am not aware of any problem  caused by applying it to all connectors, so it looks good to me. Appreciate suggestions/opinions from others :)

File: lib/migration.js [hunk 4]
```diff
@@ -437,28 +364,9 @@ function mixinMigration(Oracle) {
...
-    if (colType) {
-      return colType + (colLength ? '(' + colLength + ')' : '');
-    }
-
+  Oracle.prototype.buildColumnType = function
+  buildColumnType(propertyDefinition, colLength) {
+    var prop = propertyDefinition;
     switch (prop.type.name) {
```
Comment (jannyHou) [inline lib/migration.js]: We need to extract these code to prototype function `buildColumnType`. 
In base connector `columnDataType` calls `buildColumnType`, and expecting the specific connector to implement `buildColumnType`

File: lib/migration.js (hunk 1 above)
Comment (ssh24) [inline lib/migration.js]: Fixed in commit https://github.com/strongloop/loopback-connector-oracle/pull/110/commits/c006e7dbd1780428600bdaae3e2eb82890230110

File: lib/migration.js (hunk 4 above)
Comment (ssh24) [inline lib/migration.js]: In base connector we do have `columnDataType` and expects the child connectors to implement `buildColumnType`.

**Comment by ssh24** [comment]
@slnode test please

File: lib/migration.js [hunk 5]
```diff
@@ -437,28 +368,9 @@ function mixinMigration(Oracle) {
...
-    if (!prop) {
-      return null;
-    }
-    var colLength = columnMetadata && columnMetadata.dataLength || prop.length;
-    if (colType) {
-      return colType + (colLength ? '(' + colLength + ')' : '');
-    }
-
+  Oracle.prototype.buildColumnType = function
+  buildColumnType(propertyDefinition, colLength) {
```
Comment (loay) [inline lib/migration.js]: ``` Oracle.prototype.buildColumnType = function(propertyDefinition, colLength) {```

**Comment by loay** [comment]
@slnode test please

**Comment by loay** [comment]
@slnode test please

**Comment by loay** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by loay** [comment]
@slnode test please

**Pull request merged**